pkg/resultstore/resultstore.go

```diff
@@ -0,0 +1,152 @@
+// are created.
+type Store interface {
+	// OpenLogWriter opens a file handle abstraction for writing log lines. Logs
+	// will be automatically parsed when written and published to any active
+	// subscriptions.
+	OpenLogWriter(stepID uint64) (LogLineWriteCloser, error)
+
+	// OpenLogReader opens a file handle abstraction for reading log lines. Logs
+	// will be automatically parsed when read. Will return fs.ErrNotExist if
+	// the log file does not exist yet.
+	OpenLogReader(stepID uint64) (LogLineReadCloser, error)
+
+	// SubAllLogLines creates a new channel that streams all log lines
+	// from this result store since the beginning, and keeps on streaming new
+	// updates until unsubscribed.
+	SubAllLogLines(buffer int) (<-chan LogLine, error)
+
+	// UnsubAllLogLines unsubscribes a subscription of all status updates
+	// created via SubAllLogLines.
+	UnsubAllLogLines(ch <-chan LogLine) bool
+
+	// AddStatusUpdate adds a status update to a step. If the latest status
+	// update found for the step is the same as the new status, then this
+	// status update is skipped. Any written status update is also published to
+	// any active subscriptions.
+	AddStatusUpdate(stepID uint64, timestamp time.Time, newStatus worker.Status) error
+
+	// SubAllStatusUpdates creates a new channel that streams all status updates
+	// from this result store since the beginning, and keeps on streaming new
+	// updates until unsubscribed.
+	SubAllStatusUpdates(buffer int) (<-chan StatusUpdate, error)
+
+	// UnsubAllStatusUpdates unsubscribes a subscription of all status updates
+	// created via SubAllStatusUpdates.
+	UnsubAllStatusUpdates(ch <-chan StatusUpdate) bool
+}
+
+// LogLineWriteCloser is the interface for writing log lines and ability to
+// close the file handle.
+type LogLineWriteCloser interface {
+	io.Closer
+	// WriteLogLine will write the log line to the file and publish a parsed
+	// LogLine to any active subscriptions. An error is returned if it failed
+	// to write, such as if the file system has run out of disk space or if the
+	// file was removed.
+	WriteLogLine(line string) error
+}
+
+// LogLineReadCloser is the interface for reading log lines and ability to
+// close the file handle.
+type LogLineReadCloser interface {
+	io.Closer
+	// SetMaxLogID sets the last log ID this reader will read before artificially
+	// returning io.EOF in ReadLogLine and ReadLastLogLine. Setting this to zero,
+	// which is the default, will disable this functionality.
+	SetMaxLogID(logID uint64)
+	// ReadLogLine will read the next log line in the file, or return io.EOF
+	// when the reader has reached the end of the file.
+	ReadLogLine() (LogLine, error)
+	// ReadLastLogLine will read the entire file and return the last log line
+	// in the file. Wil return io.EOF if the file is empty.
+	ReadLastLogLine() (LogLine, error)
+}
+
+// NewStore creates a new store using a given filesystem.
+func NewStore(fs FS) Store {
+	return &store{
+		fs: fs,
+	}
+}
+
```

Comment:
If I understand you correctly, it's the latter.

The `fs FS` parameter is only used as the file system interface. If the `FS` is an `osFS` then it uses the OS's file system, which may not be empty.

Would you prefer something like `NewStoreFS` to better emphasize that it's creating a store based on the specified `FS`?

Reply:
If changing from `NewStore`, which in my head works because it should be the function to use when creating the object with minimum requirements, then I'd suggest `NewStoreFromFS` instead. `NewStoreFS` feels like it would create the file system interface to be used with the store, not the store itself.

Reply:
Conventionally you only add the suffix of the additional type that the function uses, which is clearer when there's an alternative function available without the suffix.

Ex:

- `bytes.NewBuffer` vs `.NewBufferString` https://pkg.go.dev/bytes#NewBufferString
- `ginutil.WriteUnauthorized` vs `.WriteUnauthorizedError` https://pkg.go.dev/github.com/iver-wharf/wharf-core@v1.3.0/pkg/ginutil#WriteUnauthorizedError

So `NewStoreFS` stays compliant with that convention, however it is slightly confusing due to the lack of `NewStore`.

I'm voting for keeping it as just `NewStore`, as the `New...` functions in Go are equivalent to constructors. And I like your argument there @Alexamakans of "it should be the function to use when creating the object with minimum requirements".

Reply:
I'm going to merge this. This is such a minor detail and it works well as-is, and you both have marked your reviews as approve.

If anything, I'd rather revisit the mutex/semaphore/channels implementation later. I think it got a bit convoluted. But that's for later